test_models.py automatic model discovery/runner ported from tt-torch and added to push/pr/nightly CI

kmabeeTT opened the pull request:
### Ticket
#999 

### Problem description
- We want to have the automatic tt-forge-models discovery and runner test from tt-torch running tt-xla.  It's a single test that will run any (torch for now) model from tt-forge-models repo with ability to set overrides on pcc/atol etc per test, and optionally install/rollback per-model requirements.txt if defined.

### What's changed
- Port the latest tests/runner/* files from tt-torch commit 20e10bda including test_models.py and supporting code and update slightly to work with tt-xla (Add DynamicTorchModelTester, remove CompilerConfig usage, set ComparisonConfig differently), clean up a lot of code since tt-torch version.
- Update TorchModelTester _get_forward_method_args() for list/tuple inputs (tuples for many models including albert/token_classification/pytorch-base_v2-full-eval, list for opt/bart tests like bart/pytorch-large-full-eval])
- Temporary workaround to avoid discover/import surya-ocr model since breaks torch_xla runtime with --forked after wheel refactor (issue #1166)
- Started off seeing parity with tt-torch around 240 models passing e2e. After pushing more models and updated tt-forge-models, reached ~370, few regressions, now finished with 351 models running e2e (many passing, some with reduced PCC or disabled PCC).  .test_durations file manually generated from local runs, works pretty good to balance tests.
- A whole bunch of test_config.py updates for various passing models, and skipping 55 models that are 7B+ that don't fit on device (marked bringup_status=FAILED_RUNTIME w/ reason "Too large for single chip") though these aren't run on CI currently.
- Lots of adjustments last few days after torch-xla backend was integrated with host consteval disabled, several issues opened and tests marked xfail.
- CI yml updates to include set of expected-passing 351 models tagged with `ModelStatus.EXPECTED_PASSING` in nightly (passing link below), small set of models in pr-main/push-main, along with validation of test_config.py, support in manual-test.yml.  Also run xfail + skipped tests in a separate job with fewer machines in parallel.
- CI yml QOL improvements (configurable timeout, verbosity dependent on --forked to see percentage printed while running, add 10-job parallel option, use pytest --durations=0 flag to see test times). Pull out pytest cmd and resolve it so it can be used to run --collect-only and display tests that are going to be run (useful here because tests are dynamic) for reference, and also so pytest command appears in log file with more variables resolved.
- Initial attempt at supporting both static (bringup_status + reason) results for models via test_config.py, and dynamic based on runtime results for various common situations (passing, pcc fail, compiler fail, skipped tests, etc). Added a new bringup_status=UNKNOWN because this logic is a bit best-effort for now, don't have great visibility into different types of compiler failures.  Needed to disable collection-time recording of properties with a flag.
- CI Yml Changes - Add few models to OnPush/OnPR along with test_config.py validation job run to catch typos, model name changes, see PR for more details.  Also update manual-test.yml for launching these tests.
- Removing ~14 passing duplicate torch models that are covered here.
- A whole bunch of PR feedback incorporated... more checks, cleanup code, add comments,
- Infra changes to allow static bring_status + reason to be used even for passing tests
- Probably more, see PR.

### Checklist
- [x] Lots of iteration, testing - most recent nightly before marking PR ready for review was Sept 2nd with 358 tests run e2e for each WH + BH here: https://github.com/tenstorrent/tt-xla/actions/runs/17419974014 - more after that.

kmabeeTT commented:
Lewis' torch-xla backend change went in today that this depended on, rebased and rerun everything (see passing link above, screenshot here), cleaned up lots of code.  Lots still I want to do, but in the interest of getting models running in CI sooner rather than later have to cut this off as Round 1 changes and defer any major changes to Round 2 which I'll follow up with as soon as this is merged.  Nightly takes about 2.5 hours now using 10 jobs for WH and 10 jobs for BH (BH jobs are quicker than WH). Screenshot:

<img width="1118" height="1099" alt="image" src="https://github.com/user-attachments/assets/8a08d5e4-3d81-4f6c-ad98-ee21d9729fa8" />

Screenshot of some dynamic results based on run making it to superset (queried my branch):

<img width="2062" height="208" alt="image" src="https://github.com/user-attachments/assets/139002c5-9dd3-4f2b-a74e-792853c12e2c" />

Here are some examples of various tags/bringup_status for few interesting failures that are handled.

<img width="1195" height="628" alt="image" src="https://github.com/user-attachments/assets/8dbc9fc2-5c27-46b4-9fb1-02d39d5d4b2b" />

Here is example of the percentages displayed now during running of --forked tests with one of my changes to change verbosity in call-test.yml:

```
tests/runner/test_models.py::test_all_models[dla/pytorch-dla60x-full-inference] PASSED [  2%]
tests/runner/test_models.py::test_all_models[vit/pytorch-base-full-inference] PASSED [  5%]
tests/runner/test_models.py::test_all_models[swin/image_classification/pytorch-swin_b-full-inference] PASSED [  8%]
tests/runner/test_models.py::test_all_models[qwen_1_5/causal_lm/pytorch-0_5b-full-inference] PASSED [ 11%]
```

vvukomanTT commented on .github/workflows/schedule-nightly.yml:
```diff
@@ -114,11 +140,44 @@ jobs:
+      artifact_release_run_id: ${{ needs.build-ttxla.outputs.artifacts_run_id }}
+      wheel_release_artifact_name: ${{ needs.build-ttxla.outputs.wheel_artifact_name }}
+
+  test_full_model_forge_models_not_expected_passing:
+    uses: ./.github/workflows/call-test.yml
+    secrets: inherit
+    needs: [ build-image, build-ttxla, generate-matrix ]
+    # This ensures the job runs regardless of success or failure of `nightly_tests`:
+    if: false #success() || failure()
+    with:
+      timeout_minutes: 180
+      docker_image: ${{ needs.build-image.outputs.docker-image }}
+      test_mark: 'not expected_passing'
+      test_matrix: ${{ needs.generate-matrix.outputs.test-forge-models-matrix }}
+      use-shared-runners: false # Run full model tests on Civ1
+      artifact_release_run_id: ${{ needs.build-ttxla.outputs.artifacts_run_id }}
+      wheel_release_artifact_name: ${{ needs.build-ttxla.outputs.wheel_artifact_name }}
```
I'm not sure about running the optional tests in our "main" nigthly run.
I would leave the must pass tests in the nightly and create a separate "optional" nigthly run for all the tests that are allowed to fail.
Please refer to `scheduled-optional-tests` from this [doc](https://docs.google.com/document/d/1KBGKQwXDiN8mj7c80NF9iQtAJ2RlE3y8mglSNkvX22k/edit?tab=t.0).

My main concern is that we will have a similar situation to forge-fe, where the nigthly run is always failing, not because we have core functionality that is broken but because of the "optional" tests.

Let me know what do you think about that.

kmabeeTT replied:
Yeah I do somewhat agree, I somehow dropped a commit (wtf) meant to remove this chunk from my PR and keep just the first chunk of expected passing, whoops will fix.

kmabeeTT replied:
Fixed, forgot to force push cleanup last night, whoops.

kmabeeTT commented on .github/workflows/schedule-nightly.yml:
```diff
@@ -100,12 +103,35 @@ jobs:
     needs: [ build-image, build-ttxla, generate-matrix ]
     # This ensures the job runs regardless of success or failure of `nightly_tests`:
-    if: success() || failure()
+    if: false #success() || failure()
```
This was mistake, damn I must have forgot to push a change or got something dropped, but whoops, will fix.  Was for local testing, will remove the if-false.

kmabeeTT commented on .github/workflows/schedule-nightly.yml:
```diff
@@ -49,6 +50,7 @@ jobs:
       docker_image: ${{ needs.build-image.outputs.docker-image }}
 
   nightly_tests:
+    if: false
```
Yeah this is mistake, I removed this locally last night, but somehow dropped the cleanup commit, will fix, wasn't intentional.

kmabeeTT replied:
Fixed, forgot to force push a cleanup change I made last night, done - resolved.

kmabeeTT commented on .github/workflows/call-test.yml:
```diff
@@ -153,36 +158,71 @@ jobs:
+          --splitting-algorithm least_duration"
+        echo "BASE_PYTEST_CMD=$BASE_PYTEST_CMD" >> $GITHUB_ENV
+
+    - name: Install System Deps for Forge Models Tests
+      if: ${{ matrix.build.name == 'run_forge_models_torch' }}
+      shell: bash
+      run: |
+        apt-get update
+        apt install -y libgl1 libglx-mesa0
```
So, there was a thread on that in tt-xla channel I started suggesting the same (about adding model deps into dockerfile: [link](https://tenstorrent.slack.com/archives/C08H2NCQGHJ/p1755664426856809)), but decision was not to include it in docker image but preference to install as needed when models run (longer term) or before all models in CI like above in shorter term.

vvukomanTT reviewed (approved):
Thanks for the changes Kyle, they look great 👍

kmabeeTT commented:
FYI - I just pushed a couple follow up commits that address these two TODO bullets which were important to include and also cleans up CI yml changes (more composable, DRY principle like recently shared TT Forge CI Pipelone doc mentions) FYI @vvukomanTT:

- Include some tests in onPR/onPush including pytest -vv tests/runner/test_models.py --validate-test-config to verify that test_config.py entries are all valid (useful for catching when model variant names change)
- Add custom markers support, or more markers to mark some models as wanting to be run in onPush/onPR


Here are the two changes:

```
commit 69664d31c4a11077411fb0b82a01f8fe134fd711 (HEAD -> kmabee/tt_forge_models_test_integration, origin/kmabee/tt_forge_models_test_integration)
Author: Kyle Mabee <kmabee@tenstorrent.com>
Date:   Wed Sep 3 17:57:54 2025 +0000

    Simplify/Improve/Cleanup CI yml changes for tt-forge-models tests usage

     - Reduce duplicated matrix generation (DRY!) from nighly/manual-test by moving to
       own yml file call-generate-forge-models-matrix.yml
     - It can now be used in push-main/pr-main too, to run few quick models
       for validating flow but most important validating config file, takes 5 mins
     - Add 'Validate Forge Models Tests test config file' step in call-test.yml to run
       for first matrix entry only (quick 15s job, but avoid duplication).
       Make it so tests are run even if validate step fails (still useful).
     - Update manual-test.yml to allow user specified mark for tt-forge-models
       so user can do things like "expected_passing and push" etc
     - Misc changes... rename jobs, remove redundant -q in collect tests job

commit 536271b5e082859c0f0c2d57d26f6031d18d5b3b
Author: Kyle Mabee <kmabee@tenstorrent.com>
Date:   Wed Sep 3 19:40:54 2025 +0000

    Support for and use markers in test_config.py like push, nightly, custom, etc

     - Add push marker to 4x tests mnist, mobilenetv1, gpt2, vgg19, they are
       among quickest models

```

**Some screenshot/details for more info.**

Tested new CI yml files with individual runs for pr-main, push-main, run-test, schedule-nightly, all working as expected.

Here is an example of run-test w/ tt-forge-models group with mark "expected_passing and push" for example:

<img width="481" height="891" alt="Screenshot 2025-09-03 at 5 37 38 PM" src="https://github.com/user-attachments/assets/749ba76d-29fd-43ad-9130-b296d12a0b84" />

Resulting CI run: https://github.com/tenstorrent/tt-xla/actions/runs/17445640309

Here is an on-push (push-main) CI run for example: https://github.com/tenstorrent/tt-xla/actions/runs/17453656203

Example of how validate config file [job](https://github.com/tenstorrent/tt-xla/actions/runs/17451235052/job/49556415881) looks when it catches issues with test_config.py (injected error for illustrative purposes).  It's setup to run just once on the first matrix job for tt-forge-models tests to reduce duplication.

<img width="1377" height="487" alt="image" src="https://github.com/user-attachments/assets/a885cbb8-7d28-4236-a958-25d257b1ffcc" />

kmabeeTT commented on .github/workflows/schedule-nightly.yml:
```diff
@@ -100,6 +101,13 @@ jobs:
+    secrets: inherit
+    needs: [ set-inputs ]
+    with:
+      test_group_cnt: '10' # Hardcode, lots of models in nightly.
```
Prefer to keep, but it currently takes ~2.5 hours (WH jobs are a bit longer than BH jobs) in the referenced [nightly](https://github.com/tenstorrent/tt-xla/actions/runs/17419974014) job in PR description above... 

<img width="414" height="727" alt="image" src="https://github.com/user-attachments/assets/fc2b0693-f2c8-4f69-b56b-20a3226dc491" />

I had been using 10 in tt-torch nightly for a month (would hopefulyl retire son), and tt-xla for past 3 weeks while bringing this up and had good success (lots of runners in shared pool lately).  We could cut in half from 10 to 5 and extend timeout to 6 hours for job, but that's a long time, and if anything goes wrong and need to rerun a particular group (infra error) then it's another 5+ hours to get results...

vmilosevic replied:
> how does the nightly label work to limit number of machines
We can label some of the runners as 'nightly', let's say 5 of each type, and add a label in nightly workflow to be 'n150,nightly'. That way nightly jobs will only run those 5 machines, with 10 jobs total it would take the double time.

kmabeeTT commented on tests/runner/test_utils.py:
```diff
@@ -0,0 +1,382 @@
+
+    msg = message.lower() if message else ""
+    err = (stderr or "").lower()
+    print(f"KCM Found exception: {repr(exc)} message: {msg} stderr: {err}")
```
whoops debug code, will remove.

kmabeeTT commented:
Pushed another change regarding xfail tests, and updated above section "Notes to Discuss" that previously talked about this.

I propose to tag passing tests with `EXPECTED_PASSING` status like I've been doing in tt-torch, then when a test starts failing, toggle it to `KNOWN_FAILURE_XFAIL` with hardcoded/static bringup_status + reason. They are optional, if left out, they will be dynamically filled in at runtime.

Here is some illustration of how things work today with that follow up change I pushed:

```

Option 1 : With marked xfail but no bringup_status or reason detected in test_config.py:

    "gpt_neo/causal_lm/pytorch-gpt_neo_125M-full-inference": {
        # PCC dropped issue 1234
        "status": ModelStatus.KNOWN_FAILURE_XFAIL,
    },

Get this format of XML:

  <testsuite name="pytest" errors="0" failures="0" skipped="1" tests="1" time="181.041" timestamp="2025-09-05T00:06:51.160719+00:00" hostname="aus-wh-08-special-kmabee-for-reservation-39982">
    <testcase classname="tests.runner.test_models" name="test_all_models[gpt_neo/causal_lm/pytorch-gpt_neo_125M-full-inference]" time="178.425">
      <properties>
        <property name="reason" value="PCC comparison failed. Calculated: pcc=0.9252563714981079. Required: pcc=0.99."/>
        <property name="tags" value="{'test_name': 'test_all_models', 'specific_test_case': 'test_all_models[gpt_neo/causal_lm/pytorch-gpt_neo_125M-full-inference]', 'category': 'model_test', 'model_name': 'pytorch_gpt_neo_gpt_neo_125M_nlp_causal_lm_huggingface', 'run_mode': 'inference', 'bringup_status': 'INCORRECT_RESULT'}"/>
        <property name="owner" value="tt-xla"/>
        <property name="group" value="generality"/>
      </properties>
      <skipped type="pytest.xfail" message="PCC comparison failed. Calculated: pcc=0.9252563714981079. Required: pcc=0.99."/>
    </testcase>
  </testsuite>


Option 2 : With test marked xfail but statically setting bringup_status and reason in test_config.py:

    "gpt_neo/causal_lm/pytorch-gpt_neo_125M-full-inference": {
        "reason": "PCC comparison failed. Calculated: pcc=0.9252563714981079. Required: pcc=0.99 - https://github.com/tenstorrent/tt-xla/issues/1242",
        "bringup_status": BringupStatus.INCORRECT_RESULT,
        "status": ModelStatus.KNOWN_FAILURE_XFAIL,
    },

You can see they propogate to XML:

    <testcase classname="tests.runner.test_models" name="test_all_models[gpt_neo/causal_lm/pytorch-gpt_neo_125M-full-inference]" time="226.839">
      <properties>
        <property name="reason" value="PCC comparison failed. Calculated: pcc=0.9252563714981079. Required: pcc=0.99 - [https://github.com/tenstorrent/tt-xla/issues/1242"/](https://github.com/tenstorrent/tt-xla/issues/1242%22/)>
        <property name="tags" value="{'test_name': 'test_all_models', 'specific_test_case': 'test_all_models[gpt_neo/causal_lm/pytorch-gpt_neo_125M-full-inference]', 'category': 'model_test', 'model_name': 'pytorch_gpt_neo_gpt_neo_125M_nlp_causal_lm_huggingface', 'run_mode': 'inference', 'bringup_status': 'INCORRECT_RESULT'}"/>
        <property name="owner" value="tt-xla"/>
        <property name="group" value="generality"/>
      </properties>
      <skipped type="pytest.xfail" message="PCC comparison failed. Calculated: pcc=0.9252563714981079. Required: pcc=0.99 - [https://github.com/tenstorrent/tt-xla/issues/1242"/](https://github.com/tenstorrent/tt-xla/issues/1242%22/)>
    </testcase>
```

nvukobratTT commented on venv/requirements-dev.txt:
```diff
@@ -32,9 +32,25 @@ timm
 
 wheel
+# monodepth2 model
+matplotlib
```
Should we just add these requirements per model on tt-forge-models instead of here? Seems like these are pretty localized and don't have an impact on big number of models. Therefore, should be a simpler push to get it there.

kmabeeTT replied:
Yes that is goal, but easier said than done, outside scope of this PR... so below change was definitely simpler, and in spirit of what we have in other projects until now.

Some more details on why it's not so simple:

1. Need to spend time identifying correct requirements.txt per model for it to work, and add them, and then pull in updated tt-forge-models version here (adds a dependency, every uplift of tt-forge-models bring churn... I have experienced it)
2. Not only adding requirements.txt per model, but deferring imports of packages to runtime of code rather than top level of file (see my PR here where I did it for a few models as example: https://github.com/tenstorrent/tt-forge-models/pull/85) because currently per-model requirements.txt packages are installed only before running the model, not when discovering/collecting all models during pytest parameterization searching, and so currently imports would fail due to missing packages.
3. There is currently debate (your review comments in this PR) about value of installing/uninstalling requirements.txt per model at test runtime, and perhaps relying on new infra which doesn't yet exist, so simpler to use what already exists here (this PR's goal is to just port the test_models.py and infra, technically).

AleksKnezevic commented on tests/runner/requirements.py:
```diff
@@ -0,0 +1,196 @@
+        print(msg, flush=True)
+
+
+class RequirementsManager:
```
Having dedicated requirements.txt for each model allows us to:
- Publish nice demos, telling the user what needs to be installed for each model on top of a lightweight wheel
- Support models with conflicting requirements, i.e. use transformers version that has Jax models for some tests, and a newer version for newer pytorch models in other tests. 
I thought this was a solution we settled on a while ago, wasn't it? 

Regardless on what we decide to do in the end, this works, is tested, and could be deployed to CI today bumping the supported xla model count by hundreds, and allowing people to iterate on fixing the remaining ones. If at some point we want to add a feature that we collect all the model requirements and store them in a docker, it can easily be done as a follow up PR without blocking this one.

mrakitaTT replied:
Sorry Aleks I haven't reloaded a page before publishing comments, my question was directed to Nikola :)

kmabeeTT replied:
Good discussion, though I thought we had some agreement last week and weeks prior on this.  I hope to merge this as-is, so let's move this discussion (copied replies above) to tracking ticket I just opened (https://github.com/tenstorrent/tt-xla/issues/1308) under new parent for follow up items.

nvukobratTT replied:
I assume you didn't see my last comment on this issue.. In sum, larger requests (like this one) are not a blocker from my side. I asked @kmabeeTT specifically to open a new set of issues for us to track bigger discussions and implement changes through iterations in the future. I fully agree that the focus on this PR should be porting and adjustment, rather than re-design for all of the mentioned reasons.

kmabeeTT commented on tests/runner/test_models.py:
```diff
@@ -0,0 +1,85 @@
+    [RunMode.INFERENCE],
+    ids=["inference"],
+)
+@pytest.mark.parametrize(
```
Agreed, would not neccessaily be tested on CI, but this pytest parameterization exists simply for future expansion to support op-by-op flow via this test_models.py for free (which is desired) rather than user having to spend time writing new tests to get to the bottom of the dozens of models we have today that are failing in compiler. Then, they just need to run:

```
pytest -svv tests/runner/test_models.py::test_all_models[mnist/pytorch-op_by_op-inference] |& tee mnist_op_by_op.log
```

instead of existing cmd:

```
pytest -svv tests/runner/test_models.py::test_all_models[mnist/pytorch-full-inference] |& tee mnist_e2e.log
```

I included it here so that we don't later need to do big search and replace update of all models in test_config.py and .test_durations file and so cmds I have shared in tickets recently stay valid before/after such a change is made (for this reason I also prefer to keep it as-is).

kmabeeTT commented on tests/runner/conftest.py:
```diff
@@ -0,0 +1,112 @@
+
+
+def pytest_collection_modifyitems(config, items):
+    arch = config.getoption("--arch")
```
None of these flags must be populated, arch is optional and only used to match overrides in test_config.py, for example when --arch p150 this override would be used:

```
    "swin/image_classification/pytorch-swin_b-full-inference": {
        "status": ModelStatus.EXPECTED_PASSING,
        # Was passing before, started hitting this on Aug29 : https://github.com/tenstorrent/tt-xla/issues/1168
        # AssertionError: PCC comparison failed. Calculated: pcc=0.5627762079238892. Required: pcc=0.99.
        "arch_overrides": {
            "p150": {
                "assert_pcc": False,
            },
        },
    },
```

Previous PR review comment above touches on this... I could restrict to certain set of strings and assert if something unexpected is seen though.

kmabeeTT commented on tests/runner/test_utils.py:
```diff
@@ -0,0 +1,384 @@
+        record_property("group", str(model_info.group))
+
+    # Control flow for skipped and xfailed tests is handled by pytest.
+    if test_metadata.status == ModelStatus.NOT_SUPPORTED_SKIP:
```
No crash, it's perfectly fine to have new models detected, that's kind of the entire point and by design of this that we can run them for free without any changes locally or on a CI job targeting them, they'll be reported as passing if they pass, or failed if they fail (with some of the common bringup_status's applied dynamically).  Right now I have CI jobs for expected passing tests (`-m expected_passing`) and the ones marked xfail (`-m known_failure_xfail`) but could easily set up a third (separate from main nightly to not impact nightly stability) to run the (`-m unknown`) set of tests that are new/unknown status.  Even if they crash, because --forked is used, would just cause the single test to fail, too which is nice.

kmabeeTT commented on tests/runner/test_models.py:
Thx for ideas, yeah let's discuss more later - opened issue to track follow up changes: https://github.com/tenstorrent/tt-xla/issues/1303

kmabeeTT commented on tests/runner/test_utils.py:
```diff
@@ -0,0 +1,384 @@
+            bringup_status = BringupStatus.UNKNOWN
+            reason = "Not specified"
+
+    tags = {
```
Yep I spotted this PR go in too, will rebase and make some updates (model_info, parallelization too probably).

kmabeeTT commented on .github/workflows/manual-test.yml:
```diff
@@ -14,9 +14,10 @@ on:
           - 'nightly'
           - 'model_test'
+          - 'tt_forge_models'
           - 'Custom'
```
I thought so too, but it was there before I did anything... sure, will update.

kmabeeTT commented on .github/workflows/call-test.yml:
```diff
@@ -153,36 +158,82 @@ jobs:
-        if [[ "${{ matrix.build.name }}" == "run_torch" || "${{ matrix.build.name }}" == "run_large_jax_models" ]]; then
-          PYTEST_FORKED="--forked"
+        PYTEST_FORKED=""
+        for n in run_torch run_large_jax_models run_forge_models_torch; do
```
I found that you guys had this in tt-xla already, and --forked was added here originally as protection against large models that may segfault or crash and take down entire CI group of tests... with --forked they are in seperate processes per test... it has helped (quite a lot) for me the last 3 weeks to use it too, especially when running such a large number of models where the risk of anything going wrong is heightened.  In tt-torch we don't use it, and I found myself having to rerun failing jobs occasionally.  

Downside: there is a real runtime performance impact though I have noticed... run 2 variants of a model in --forked and they don't share any state and take some time, but if run in the same process, I have found subsequent related models be faster.

tt-forge-fe has a different solution that DevEx folks will port where pytests are run in a wrapper script in subprocess, and if a crash is encoutnered, it will spawn a new process to keep running the rest (or optionally, start over from beginning of list with the offending test removed)... this flow actually enjoys the cached/speedup seen without using --forked in my experiments the last few weeks.

nvukobratTT replied:
Yup, Crash Detector, we had solved this kind of problem. It's best to ping @nsumrakTT to make similar kinds of changes on XLA as well :)))

kmabeeTT commented on tests/runner/conftest.py:
```diff
@@ -0,0 +1,112 @@
+
+
+def pytest_addoption(parser):
+    parser.addoption(
```
To help close on open Q's and get this merged sooner, I went ahead and made change for this. Added this:

```
# Allowed architecture identifiers for arch_overrides and --arch option
ALLOWED_ARCHES = {"n150", "p150"}
```
If violated, user sees this:

```
pytest: error: argument --arch: invalid choice: 'p300' (choose from 'n150', 'p150')
```

And enhanced the check with `--validate-test-config` which is a separate check in CI, can be run locally to catch if `test_config.py` entries refer to invalid arch, example:

```
ERROR: Found invalid arch_overrides entries (unknown arches):
  - swin/image_classification/pytorch-swin_t-full-inference: unknown arch 'p450'

Allowed arches: n150, p150
```

Will push later today along with bunch of other small tweaks.

nvukobratTT replied:
Nice! Also, should we specify other available archs as well? E.g., do we have any limitations with running these models on the multi-chip setup (DP, TP, PP)?

kmabeeTT commented on tests/runner/test_config.py:
```diff
@@ -0,0 +1,1815 @@
+# SPDX-FileCopyrightText: (c) 2025 Tenstorrent AI ULC
```
Yes this is valid concern - let's discuss more as next steps. Opened issue to track: https://github.com/tenstorrent/tt-xla/issues/1305 ..

Though immediate goal for this PR is to port infra that already exists in tt-torch (improved/updated a bit for tt-xla, but nothing major) for parity tracking and to run the reasonable set of ~500 models in CI.

Using SQL DB to track/manage model status is interesting idea, it brings some complexities and challenges though. One thing I'm concerned with is not wanting to lose a history (per commit) of model status if we only maintain latest-and-greatest results in DB, because it's important to know what status of models is at a given commit (being able to sync back, or work on branches) and a checked in file in github perfectly preserves that relationship today... having an external thing to track adds some uncertainty to relationship unless controlled well.

FWIW, current changes can:

1.  Already run new models without any test_config.py update (just need to add a CI job to run them (-m unknown), will come in follow up PR) 
2. `test_config.py` already supports and uses tagging for models between push, nightly, weekly (push is used, all else defaults to nightly right now). 
3. prevent new/failing models from impacting already passing tests (run in nightly) by statically tracking them in `test_config.py` since no other easy method exists right now.

Preference is to merge this and discuss long term strategy next. FYI @AleksKnezevic

kmabeeTT commented on tests/runner/conftest.py:
```diff
@@ -0,0 +1,112 @@
+            try:
+                item.add_marker(getattr(pytest.mark, marker_name))
+            except Exception:
+                # Silently ignore unknown markers; they should be declared in pytest.ini
```
Sorry looks like this isn't happening like I thought and try/except can be removed after some experiments, chatGPT confirms:

> In pytest, pytest.mark.<anything> never raises. It always returns a MarkDecorator object, even if the marker wasn’t declared in pytest.ini.

For context, I thought this was needed for feature I added to support user-defined / custom markers in `test_config.py` without needing to add them to pytest.ini.  With this a user who wants to target specific group of tests affected by a bug (eg. when testing against a fix), they can do so.  For one-off experiments, simply add a marker into the bunch of test entries you care about locally:

```
    "mnist/pytorch-full-inference": {
        "status": ModelStatus.EXPECTED_PASSING,
        "markers": ["issue_1234"],
    },
    "mobilenetv1/pytorch-mobilenet_v1-full-inference": {
        "status": ModelStatus.EXPECTED_PASSING,
        "markers": ["issue_1234"],
    },
```

Which get translated to code above you saw, then just run:

```
pytest -svv tests/runner/test_models.py -m issue_1234 |& tee kyle_test_issue_1234_fix.log
```

No exception thrown above, will remove try/except.

kmabeeTT commented on tests/runner/conftest.py:
```diff
@@ -0,0 +1,112 @@
+    return meta
+
+
+def pytest_collection_modifyitems(config, items):
```
Done, added comments above each function in tests/runner/conftest.py and tests/runner/test_utils.py for this, will push along with other updates later today.

kmabeeTT commented on .github/workflows/call-generate-forge-models-matrix.yml:
```diff
@@ -0,0 +1,50 @@
+        run: |
+            SCRIPT_PATH=".github/scripts/generate_test_model_matrix.py"
+            TEST_MATRIX_PATH=".github/workflows/test-forge-models-matrix.json"
+            TESTS_TO_PARALLELIZE='[
```
Looks like No - not exactly today, would require updates to `.github/scripts/generate_test_model_matrix.py` which is script used by existing nightly tests added by Andrej a while back,  that I just copied how they worked.  It relies on matching "name" and "runs-on" per item with .json file by design, though looks like "dir" can be excluded here though, so I'll do that for now.

vvukomanTT replied:
That wouldn't be possible as the script `.github/scripts/generate_test_model_matrix.py` also manipulates `test-full-model-matrix.json` which contains non-unique test suite names.
e.g.
```
{ "runs-on": "n150",        "name": "run_jax",           "dir": "./tests/jax/single_chip", "extra_test_marks": "not large" },
{ "runs-on": "p150",        "name": "run_jax",           "dir": "./tests/jax/single_chip", "extra_test_marks": "not large" },
{ "runs-on": "n300",        "name": "run_jax",           "dir": "./tests/jax/multi_chip/n300" },
 ```

kmabeeTT commented on .github/workflows/manual-test.yml:
```diff
@@ -159,3 +182,17 @@ jobs:
       wheel_release_artifact_name: ${{ needs.build-ttxla.outputs.wheel_artifact_name }}
+
+  test_tt_forge_models:
+    uses: ./.github/workflows/call-test.yml
+    if: inputs.preset == 'tt_forge_models' && (success() || failure())
+    secrets: inherit
+    needs: [ build-image, test-setup, build-ttxla, generate-forge-models-matrix ]
```
Good point, fixed.

kmabeeTT commented on .github/workflows/manual-test.yml:
```diff
@@ -14,9 +14,10 @@ on:
           - 'Custom'
       test_mark:
-        description: 'Test mark to run (if preset is Custom)'
+        description: 'Test mark to run (if preset is Custom or tt_forge_models'
```
Whoops thx, fixed will push with big update soon.

kmabeeTT commented on tests/runner/test_utils.py:
```diff
@@ -0,0 +1,384 @@
+from tests.utils import BringupStatus, Category
+
+
+class ModelStatus(Enum):
```
Sure, this was big update but I went ahead with rename, will push alongside other feedback soon.

kmabeeTT commented on tests/runner/requirements.py:
```diff
@@ -0,0 +1,196 @@
+        if not self.requirements_path:
+            return self
+
+        if os.environ.get("TT_TORCH_DISABLE_MODEL_REQS", "0") == "1":
+            return self
```
whoops yes, want to keep it, updated to use TT_XLA_*.

kmabeeTT commented on tests/runner/requirements.py:
```diff
@@ -0,0 +1,196 @@
+
+        # Acquire a global lock for pip operations
+        lock_path = os.path.join(
+            tempfile.gettempdir(), "tt_torch_model_requirements.lock"
```
whoops, fixed, thx.

kmabeeTT commented on tests/runner/test_utils.py:
```diff
@@ -0,0 +1,384 @@
+    def _get_input_activations(self):
+        return self._load_inputs()
+
+    def _get_forward_method_args(self):
+        return super()._get_forward_method_args()
+
+    def _get_forward_method_kwargs(self):
+        return super()._get_forward_method_kwargs()
```
Good point, removed.

kmabeeTT commented on tests/runner/test_utils.py:
```diff
@@ -0,0 +1,384 @@
+
+    # --- TorchModelTester interface implementations ---
+
+    def _get_model(self):
+        return self._load_model()
+
+    def _get_input_activations(self):
+        return self._load_inputs()
```
Good call, updated with both suggestions.

```
    @override
    def _get_model(self):
        sig = inspect.signature(self.loader.load_model)
        if "dtype_override" in sig.parameters:
            return self.loader.load_model(dtype_override=torch.bfloat16)
        return self.loader.load_model()

    @override
    def _get_input_activations(self):
        sig = inspect.signature(self.loader.load_inputs)
        if "dtype_override" in sig.parameters:
            return self.loader.load_inputs(dtype_override=torch.bfloat16)
        return self.loader.load_inputs()
```

kmabeeTT commented on tests/runner/test_utils.py:
```diff
@@ -0,0 +1,384 @@
+    # Store variant tuple along with the ModelLoader
+    for loader_path, variant_tuples in loader_paths.items():
+        for variant_info in variant_tuples:
+            test_entries.append({"path": loader_path, "variant_info": variant_info})
```
Good call, now usage in test_models.py is safer:

```
    loader_path = test_entry.path
    variant, ModelLoader = test_entry.variant_info
```

nvukobratTT commented on tests/runner/test_config.py:
```diff
@@ -0,0 +1,1815 @@
+        "status": ModelStatus.EXPECTED_PASSING,
+    },
+    "gpt_neo/causal_lm/pytorch-gpt_neo_1_3B-full-inference": {
+        "required_pcc": 0.98,
```
I'm fine with temporarily disabling PCC for some models, just to keep tracking their functionality at least, while debugging issues. Ofc, these issues should be of higher priority. If that isn't possible, we should just xfail the test. 

@mrakitaTT, the good part of having test configurations as part of code is that we can, with no effort, check how long certain configs are out there. In case of a disabled PCC, if it's more than a few days without an active issue that is WIP, we should probably just xfail it and try to bisect it (or patch it) later. WDYT? 

I had similar reasoning on FFE regarding PCC. E.g., for slightly lower models, we reduced PCC from 0.99 to 0.97. Depending on the model, the evaluation would still be pretty good, so I didn't want to block it from having a functional model. Ofc, if we're not aligned on this one we should discuss as well

AleksKnezevic replied:
@nvukobratTT, don't worry, it won't be a blocker since this is already merged 😄. I completely agree about automation. The more models/architectures we have the more critical this will be. Did you have someone on FFE that did this kind of automation work? Do you have any proposals for how to change this process? @kmabeeTT is out this week, but let's have a discussion about it when he comes back.

nvukobratTT replied:
>@nvukobratTT, don't worry, it won't be a blocker since this is already merged 😄. I completely agree about automation.  

Blocker in terms of correcting the diff between the proposal and the current test_config file (if any) 😂

> Did you have someone on FFE that did this kind of automation work? Do you have any proposals for how to change this process? @kmabeeTT is out this week, but let's have a discussion about it when he comes back.

Yes, we had a few iterations over this one, and we might be able to port a similar approach here as we discussed initially. Let's discuss once @kmabeeTT is back to have a full group on board and agree on an approach. 

Once we draft a plan, I can put an engineer who pushed most of this automation on FFE :))

nvukobratTT commented:
> Hi @nvukobratTT / @mrakitaTT plz re-review/approve for me to merge this PR for initial infra + torch models in CI and set the stage for easier future follow ups. Updates since Friday:
> 
> 1. I implemented all feedback for the things that were agreed upon and doable in short term, plz take a look
> 2. Opened parent ticket [test_models.py automatic model discovery and runner improvements tracking issue #1307](https://github.com/tenstorrent/tt-xla/issues/1307) and 3 sub items for follow up discussions for threads above that don't seem quickly resolvable.
> 3. Rebased on latest Sept 6th (Sept 7 tree is broken), few models regressed, bisected, opened tickets, updated test_config.py as needed, and rerun all models on CI again here, passing link: https://github.com/tenstorrent/tt-xla/actions/runs/17534315633
> 
> If we like the idea to move all badly failing models to PCC checking enabled but marked xfail instead of expected_passing, then I'll wrap this up with one more commit before merging to do that for the group of tests in test_config.py (talked about this in a previous message I posted in here [above](https://github.com/tenstorrent/tt-xla/pull/982#issuecomment-3256815503)).
> 
> I spent quite a while looking at superset reporting (adjusting queries to target my branch) to try and increase confidence that no surprises once this is released, and there is one thing open I found today now that I recently learned about tt-xla dashboard: The XML contain "`test_name`" and "`specific_test_name`" which are populated correctly... see screenshot. Problem is, the Pie Chart on tt-xla dashboard filters by unique test_name rather than specific_test_name and because all models are under "test_models" we only get 1 test per bringup_status reported. Need to change that query somehow, or as hack/workaround/fix I can make one line change to output same field (containing model name, variant name) as test_name (ie same value for test_name and specific_test_name).
> 
> <img alt="image" width="2000" height="681" src="https://private-user-images.githubusercontent.com/118925087/486585755-787ca797-5622-4383-891e-81b2f604ca6c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTczMjQ4MTEsIm5iZiI6MTc1NzMyNDUxMSwicGF0aCI6Ii8xMTg5MjUwODcvNDg2NTg1NzU1LTc4N2NhNzk3LTU2MjItNDM4My04OTFlLTgxYjJmNjA0Y2E2Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwOFQwOTQxNTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02MjhlYzBmYjc5ZWJkOTMwNTMzYzg4MWRjMmY0YjAyZTBlMDZhODIxMzYwNGQzZjZlYmEyNmFhNTcyZmY4NWE4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.sXqjpia1q9OEm5mKrmVLsfO0eS9B73Tgl1FIkuXKLwk">
> Here is what I mean:
> 
> <img alt="image" width="1423" height="425" src="https://private-user-images.githubusercontent.com/118925087/486586175-8678c558-1918-4dac-a005-44aa06219cc9.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTczMjQ4MTEsIm5iZiI6MTc1NzMyNDUxMSwicGF0aCI6Ii8xMTg5MjUwODcvNDg2NTg2MTc1LTg2NzhjNTU4LTE5MTgtNGRhYy1hMDA1LTQ0YWEwNjIxOWNjOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwOFQwOTQxNTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jYzE3Y2U3NjViYTNhYjdjZTQwMzBmMGFhNzk5YzliZmZjYTc2MTMyMTg0MWFkZGIxMmFhOGM5NWI1MGE2YmQzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.itJ4O2quiaY5jk5BMe8yrcyEc93LYxIlbAxrkcXehU4">
> Tag @tt-mpantic for this question on preference.

Thanks @kmabeeTT! This is a great push and effort! I'm fine with merging this PR as is. Let's have a follow-up discussion on the related topic for further improvements. 

Regarding the superset comment, if current queries don't work, we could just create a new superset dataset for this purpose, and a new set of charts until we stabilize and unify this logic with the existing XLA infra. As mentioned, let's sync on these topics over the Slack :))

kmabeeTT commented:
Thanks for follow up reviews / feedback - Planning to merge this imminently before tt-xla tree gets busy again since I am rebased and tested on the TOTT and finally just finished working around some issues.  Bunch of updates from today, including few surprises I had to push past:

1. Rebased on latest, resolved conflicts, pushed few more commits on top.
2. Removing ~14 duplicate torch models. Went through them all, I could not rm -rf entire torch test folder mainly because I am only running passing models (or models that were reecntly passing) and do not yet have listed in test_config.py the dozens of models that are failing in compiler, but I removed all that were covered.
3. 56 models failing with `assert isinstance(a, FakeTensor)` assert, bisected due to torch/torchvision changes with "preinstall requirements.txt" change, I opened issue https://github.com/tenstorrent/tt-xla/issues/1324 to track, and submitted workaround for now to not use the pre-installed requirements.txt becase would have been really annoying to remove the 56 models.  Needs debug.
4. A tt-github-actions change was needed to propagate reason->test.error_message->DB for tests not pytest.skip or pytest.xfailed where we want to specify a reason and have it show up in DB (PCC failing tests with static bringup_status + reason), took a while to test, PR here for: https://github.com/tenstorrent/tt-github-actions/pull/77 tagged @vmilosevic 
5. There is also an issue with artifacts from `test_forge_models_xfail` [job](https://github.com/tenstorrent/tt-xla/actions/runs/17571375251/job/49908419634) were not being uploaded to DB due to space in artifact name via "-m known_failure_xfail or not_supported_skip"  , root caused and opened issue here https://github.com/tenstorrent/tt-github-actions/issues/78, and applied workaround in this PR `call-test.yml` in the meantime to sanitize inputs.test_mark before using in name of artifacts to be uploaded. Can be undone once tt-github-actions fixed.
6. CODEOWNERS - updated as per above.

Need to update description of this PR with few more recent details, and will click auto-merge in a few. OnPR before I pushed CODEOWNERS update is looking [good](https://github.com/tenstorrent/tt-xla/actions/runs/17572431528/job/49911331858?pr=982).
